pytest_zigzag/__init__.py

```diff
@@ -93,29 +95,78 @@ def _get_ci_environment(session):
+    white_list = ['asc', 'mk8s']
+    if not any(x == highest_precedence for x in white_list):
+        raise RuntimeError(
+            "The value {} is not a valid value for the 'ci-environment' configuration".format(highest_precedence))
+
+    return highest_precedence
+
+
+def _get_option_of_highest_precedence(session, option_name):
+    """looks in the session and returns the option of the highest precedence
+    This assumes that there are options and flags that are equivalent
+
+    Args:
+        session (_pytest.main.Session): The pytest session object
+        option_name (str): The name of the option
+
+    Returns:
+        str: The value of the option that is of highest precedence
+        None: no value is present
+    """
     #  Try to get configs from CLI and ini
     try:
-        cli_option = session.config.getoption('--ci-environment')
+        cli_option = session.config.getoption("--{}".format(option_name))
     except ValueError:
         cli_option = None
     try:
-        ini_option = session.config.getini('ci-environment')
+        ini_option = session.config.getini(option_name)
     except ValueError:
         ini_option = None
-
-    # Determine if the option passed with the highest precedence is a valid option
-    highest_precedence = cli_option or ini_option or 'asc'
-    white_list = ['asc', 'mk8s']
-    if not any(x == highest_precedence for x in white_list):
-        raise RuntimeError(
-            "The value {} is not a valid value for the 'ci-environment' configuration".format(highest_precedence))
-
+    highest_precedence = cli_option or ini_option
     return highest_precedence
 
 
 # ======================================================================================================================
 # Hooks
 # ======================================================================================================================
+@pytest.hookimpl(trylast=True)
+def pytest_sessionfinish(session, exitstatus):
+    """This hook is used by pytest to build the junit XML
+    Using ZigZag as a library we upload in the pytest runtime
+
+    Args:
+        session (_pytest.main.Session): The pytest session object
+        exitstatus
+    """
+    SESSION_MESSAGES.drain()  # need to reset this on every pass through this hook
+    if session.config.pluginmanager.hasplugin('junitxml'):
+        zz_option = _get_option_of_highest_precedence(session, 'zigzag')
+        if zz_option:
+            qtest_project_id = _get_option_of_highest_precedence(session, 'qtest-project-id')
+            junit_file_path = getattr(session.config, '_xml', None).logfile
+            if junit_file_path:  # extra test to make double sure we have a file path recommended by Ryan
+                try:
+                    if not qtest_project_id:
+                        raise RuntimeError("'qtest-project-id' is required when using ZigZag")
+                    zz = ZigZag(junit_file_path, os.environ['QTEST_API_TOKEN'], qtest_project_id)
```

Comment:
This will never work because the signature is wrong. ZigZg requires the following:

```python
        """ Create a ZigZag facade class object. The ZigZag class uses the Facade pattern to call out to
        subsystems and sub Facades.

        Args:
            junit_xml_file_path (str): A file path to a XML element representing a JUnit style testsuite response.
            qtest_api_token (str): Token to use for authorization to the qTest API.
            qtest_project_id (int): The target qTest project for the test results.
            qtest_test_cycle (str): The parent qTest test cycle for test results.
                (e.g. Product Release codename "Queens")
            pprint_on_fail (bool): A flag for enabling debug pretty print on schema failure.
        """
```

Reply:
Fixed